Add MkTemp

arp242 opened the pull request:
This ensures a sane creation of temporary files.

e.g. to fix this error:
https://sentry.io/teamwork/desk/issues/298173930/

arp242 committed:
Add MkTemp

This ensures a sane creation of temporary files.

e.g. to fix this error:
https://sentry.io/teamwork/desk/issues/298173930/

JaTochNietDan commented on tmpdir.go:
```diff
@@ -26,3 +28,69 @@ func Setup(path string) (err error) {
+// Subdirectories are not supported. The Setup() function must be called first.
+func MkTemp(filename string) (*os.File, error) {
+	if Dir == "" {
+		panic("Dir is empty. Setup() needs to be called first.")
```
Should we not just return an error here?

arp242 replied:
Possibly? It's actually from when it just returned a `string` and didn't return an error.

But it's a clear programming/state error, and seems to be an appropriate use of panic to me.

ready4god2513 replied:
Yeah, that's the same pattern we use with `Must`.  I think it makes sense.

ready4god2513 replied:
Not a recoverable error

flimzy replied:
It may not be recoverable, but it can be handled more gracefully than a panic.

My 2¢:  Use `MkTemp` for things that are done during requests, etc, and `MustMkTemp` for things that happen during start-up, etc.

But it's a bikeshedding discussion, so I don't really care :P

arp242 replied:
> MustMkTemp for things that happen during start-up, etc.

I don't know if we ever use this; I think not. I just changed it to an error now. We can always add `Must*()` later of we actually want to use it.

arp242 committed:
Return error

arp242 committed:
Provide Windows stub for isEEXISTS()

Some people use Windows, so this makes sure the package doesn't error
out for them.

It's a non-essential feature, so this is okay for now.

cesartw commented on tmpdir.go:
```diff
@@ -26,3 +28,69 @@ func Setup(path string) (err error) {
+//
+// Subdirectories are not supported. The Setup() function must be called first.
+func MkTemp(filename string) (*os.File, error) {
+	if Dir == "" {
```
Any good reason to use the global namespace? I've seen this as a common practice but introduced a few issues:
1. Test cases may interfere with each other
2. Read/write within goroutines is not safe
3. Nobody has control over it.
4. Cannot reuse the package at runtime(safely anyway)
5. package is `broken` from the get-go

If there's an actual need for this, we may as well look into a singleton(https://golang.org/pkg/sync/#Once.Do)

ready4god2513 replied:
The primary use case of this package is to load at startup and then write files to it.  So it wouldn't be initialized more than once.  But I do agree it probably shouldn't be public or should be protected with sync.

arp242 replied:
The idea is that there's one global "temp directory". The reason is like this is because ... this is how it was written in Desk :-)

Since we didn't have any helper functions before this PR, exposing `Dir` was needed. I'm not against changing that, for example by unexporting `Dir`, but doing it is a bit outside the scope of this PR, I think.

arp242 replied:
I'll look at this after this PR is merged btw
